lib/babe/epoch_handler.go

```diff
@@ -115,6 +122,14 @@ func (h *epochHandler) run(ctx context.Context, errCh chan<- error) {
 		case <-ctx.Done():
 			return
 		case <-swt.timer.C:
+			// we must do a time correction as the slot timer sometimes is triggered
+			// before the time defined in the constructor due to an inconsistency
+			// of the language -> https://github.com/golang/go/issues/17696
```

Comment:
I'm still skeptical on this; I ran the following test (adding load and running a bunch of timers) and none occur before the expected end time... Can you try running it on your end?

```go

func Test_Timers(t *testing.T) {
	// Add CPU load
	ctx, cancel := context.WithCancel(context.Background())
	wg := new(sync.WaitGroup)
	numGoroutines := runtime.NumCPU()
	wg.Add(numGoroutines)
	for i := 0; i < numGoroutines; i++ {
		go func() {
			defer wg.Done()
			i := 0
			b := make([]byte, 1024)
			for {
				for j := range b {
					b[j] = 2*b[j] ^ 0b1111_1111 + 1
				}

				i++
				if i%1000 == 0 && ctx.Err() != nil {
					return
				}
			}
		}()
	}

	type timerData struct {
		startTime time.Time
		timer     *time.Timer
	}
	const timeout = 2 * time.Second
	const sleepBetween = 2 * time.Millisecond
	const numberOfTimers = 100
	timers := make([]timerData, numberOfTimers)

	for i := 0; i < numberOfTimers; i++ {
		timers[i].startTime = time.Now()
		timers[i].timer = time.NewTimer(timeout)
		time.Sleep(sleepBetween)
	}

	for i := 0; i < numberOfTimers; i++ {
		expectedEndTime := timers[i].startTime.Add(timeout)
		received := <-timers[i].timer.C
		endTime := time.Now()
		if endTime.Before(expectedEndTime) {
			t.Errorf("expected end time %s to be after %s (received %s from channel)",
				endTime, expectedEndTime, received)
		}
	}

	cancel()
	wg.Wait()
}

```

Reply:
I tested this code many times using `go clean -testcache && go test -run ^Test_Timers$ github.com/ChainSafe/gossamer/lib/babe -v` and it did not fail